esp-hal/src/lib.rs

```diff
@@ -660,6 +662,32 @@ pub(crate) fn lock<T>(state: &LockState, f: impl FnOnce() -> T) -> T {
     }
 }
 
+/// Data protected by [LockState]
+#[allow(unused)]
+pub(crate) struct Locked<T> {
+    lock_state: LockState,
+    data: UnsafeCell<T>,
+}
+
+#[allow(unused)]
+impl<T> Locked<T> {
+    /// Create a new instance
+    pub(crate) const fn new(data: T) -> Self {
+        Self {
+            lock_state: LockState::new(),
+            data: UnsafeCell::new(data),
+        }
+    }
+
+    /// Provide exclusive access to the protected data to the given closure
+    pub(crate) fn with<R>(&self, f: impl FnOnce(&mut T) -> R) -> R {
```

Comment:
With this API you can call `with` inside `with` and `LockState` is not reentrant, so we'll get a panic if we do that. I suppose it's an internal API so maybe it's okay to live with this limitation.